src/gradient_descent.c

```diff
@@ -0,0 +1,380 @@
+                           gradient_real initialPoint[],
+                           gradient_real direction[], int n,
+                           gradient_real tol) {
+
+  // Distance moved
+  gradient_real distance = 0.0;
+
+  // Distance moved previously
+  gradient_real previousDistance = 0.0;
+
+  // Minimum is bracketed between a and b
+  gradient_real a = (bracket.a < bracket.c ? bracket.a : bracket.c);
+  gradient_real b = (bracket.a > bracket.c ? bracket.a : bracket.c);
+
+  // xMin : Point with the smallest function value found so far
+  // secondXMin : Point with the second smallest function value found so far
+  // prevSecondXMin : Previous value of secondXMin
+  gradient_real prevSecondXMin, secondXMin;
+  *xMin = prevSecondXMin = secondXMin = bracket.b;
+
+  // Value of the function at the above points
+  gradient_real fXMin, fPrevSecondXMin, fSecondXMin;
+  fXMin = fPrevSecondXMin = fSecondXMin =
+      (*func)(f, *xMin, initialPoint, direction, n);
+
+  for (int iter = 0; iter < ITMAX_BRENT; iter++) {
+    gradient_real midpoint = 0.5 * (a + b);
+
+    // Tolerance for estimates
+    gradient_real tol1 = tol * fabs(*xMin) + EPS;
+    gradient_real tol2 = 2.0 * tol1;
+
+    // Checks if we found the minimum with enough precision
+    if (fabs(*xMin - midpoint) <= (tol2 - 0.5 * (b - a))) {
+      *xMin = *xMin;
+      return fXMin;
+    }
+
+    // No need to evaluate function if we haven't moved by at least tol1
+    if (fabs(previousDistance) > tol1) {
+      // Construct trial parabolic fit
+      gradient_real r = (*xMin - secondXMin) * (fXMin - fPrevSecondXMin);
+      gradient_real q = (*xMin - prevSecondXMin) * (fXMin - fSecondXMin);
+      gradient_real p = (*xMin - prevSecondXMin) * q - (*xMin - secondXMin) * r;
+      q = 2.0 * (q - r);
+
+      if (q > 0.0) {
+        p = -p;
+      }
+      q = fabs(q);
+
+      // Check the acceptability of parabolic fit
+      if (fabs(p) >= fabs(0.5 * q * previousDistance) || p <= q * (a - *xMin) ||
+          p >= q * (b - *xMin)) {
+        // Take the golden section step
+        previousDistance = *xMin >= midpoint ? a - *xMin : b - *xMin;
+        distance = CGOLD * previousDistance;
+      } else {
+        // Take the parabolic step
+        previousDistance = distance;
+        distance = p / q;
+        gradient_real u = *xMin + distance;
+        if (u - a < tol2 || b - u < tol2) {
+          distance = SIGN(tol1, midpoint - *xMin);
+        }
+      }
+    } else {
+      // Take the golden section step
+      previousDistance = *xMin >= midpoint ? a - *xMin : b - *xMin;
+      distance = CGOLD * previousDistance;
+    }
+
+    // current: Most recent point at which function was evaluated
+    gradient_real current = fabs(distance) >= tol1
+                                ? *xMin + distance
+                                : *xMin + SIGN(tol1, distance);
+    gradient_real fCurrent = (*func)(f, current, initialPoint, direction, n);
```

Comment:
You can directly use thee function instead of de-referencing it . Please see above comment